frame/support/src/evm.rs

```diff
@@ -0,0 +1,122 @@
+// This file is part of Darwinia.
+//
+// Copyright (C) 2018-2021 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// --- crates.io ---
+use ethereum::TransactionMessage;
+use sha3::{Digest, Keccak256};
+// --- darwinia-network ---
+use ethereum_primitives::{H160, H256, U256};
+// --- paritytech ---
+use codec::{Decode, Encode};
+use frame_support::PalletId;
+use sp_std::vec;
+use sp_std::vec::Vec;
+
+pub const POW_9: u32 = 1_000_000_000;
+pub const INTERNAL_TX_GAS_LIMIT: u32 = 300_000_000;
+pub const SELECTOR: usize = 4;
+pub const TRANSFER_ADDR: &'static str = "0x0000000000000000000000000000000000000015";
+
+pub trait IntoDvmAddress {
+	fn into_dvm_address(&self) -> H160;
+}
+
+#[derive(Clone, Copy, Eq, PartialEq, Encode, Decode)]
+pub struct ContractId(pub [u8; 8]);
+
+/// Convert from contract id to dvm address
+impl IntoDvmAddress for ContractId {
+	fn into_dvm_address(&self) -> H160 {
+		let mut bytes = vec![0u8; 12];
+		bytes.append(&mut self.0.to_vec());
+		H160::from_slice(&bytes)
+	}
+}
```

Comment:
Suggest to not adding new Types since PalletId can meet the requirements

Reply:
In my mind, `PalletId` is designed to generate a special module account id for pallets to identify a unique pallet account.

What we need is to generate a contract id instead of a pallet module id. Thus, I think directly use `PalletId` in runtime config might be confusing to others. Try to imagine that a pallet with two config type classes is PalletId.

Reply:
As the name hints, let's define the PalletId as a unique ID with type `[u8; 8]` representing this pallet, substrate account and DVM account are generated from the same pallet id.

Adding extra types add more concept and complexity, prefer to keep it simple